### PR TITLE
refactor(perception_utils): refactor perception_utils

### PR DESCRIPTION
**Pull request by scepter914**
## Description

This PR separate files to make similar to tier4_autoware_util.
This PR changes as below.

- Move the function in perception_utils.hpp to conversion.hpp, matching.hpp, and transform.hpp.
- Add functions to object_classification.hpp to adjust existing interface.
- Adjust test function as changing point.
- Refactor overall including copyright.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/perception_utils/include/perception_utils/matching.hpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#include "tier4_autoware_utils/geometry/boost_polygon_utils.hpp"
+
+#include <boost/geometry.hpp>
+
+#include <vector>
+
+namespace perception_utils
+{
+template <class T1, class T2>
+inline double get2dIoU(const T1 source_object, const T2 target_object)
```
Comment (kenji-miyake) [inline common/perception_utils/include/perception_utils/matching.hpp:29]: I think `inline` is redundant for templates (except for specialization).

```suggestion
double get2dIoU(const T1 source_object, const T2 target_object)
```

Comment (scepter914) [reply common/perception_utils/include/perception_utils/matching.hpp]: fixed at [ca81de8](https://github.com/autowarefoundation/autoware.universe/pull/1954/commits/ca81de8a897e2e3c1a398bfc8edd03a77804d4cd)

File: common/perception_utils/include/perception_utils/conversion.hpp [hunk 2]
```diff
@@ -12,25 +12,15 @@
...
-    if (highest_prob < _class.probability) {
-      highest_prob = _class.probability;
-      label = _class.label;
-    }
-  }
-  return label;
-}
-
-autoware_auto_perception_msgs::msg::DetectedObject toDetectedObject(
+inline autoware_auto_perception_msgs::msg::DetectedObject toDetectedObject(
```
Comment (kenji-miyake) [inline common/perception_utils/include/perception_utils/conversion.hpp:23]: By the way, why did you move functions to `.hpp`?
I think it's not recommended, it's just allowed for work time efficiency.
So you don't need to move all functions to header files.

Comment (scepter914) [reply common/perception_utils/include/perception_utils/conversion.hpp]: Thank you for thoughtful review.
I intend to imitate to tier4_autoware_util composition, but I agree with your recomendation.
So I fixed at [c2f9caf](https://github.com/autowarefoundation/autoware.universe/pull/1954/commits/c2f9cafae0a58f8391d7e42aa5debebd17dd0f22).

File: common/perception_utils/include/perception_utils/transform.hpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+#endif
+
+#include "boost/optional.hpp"
+
+#include <string>
+
+namespace
+{
+[[maybe_unused]] boost::optional<geometry_msgs::msg::Transform> getTransform(
```
[the remark this answers is not in the document]
Comment (scepter914) [reply common/perception_utils/include/perception_utils/transform.hpp:35]: I don't know this intention.
@purewater0901 @takayuki5168 Would you please reply this comment?

File: common/perception_utils/include/perception_utils/conversion.hpp (hunk 2 above)
Comment (kenji-miyake) [reply common/perception_utils/include/perception_utils/conversion.hpp]: Regarding `tier4_autoware_util`, I've shortened the process previously because it was in the prototyping phase.
I'm sorry for confusing you.

And thank you for fixing it.

File: common/perception_utils/include/perception_utils/object_classification.hpp [hunk 4]
```diff
@@ -17,11 +17,27 @@
...
+  const std::vector<ObjectClassification> & object_classifications)
+{
+  std::uint8_t label = ObjectClassification::UNKNOWN;
+  float highest_prob = 0.0;
+  for (const auto & object_classification : object_classifications) {
+    if (highest_prob < object_classification.probability) {
+      highest_prob = object_classification.probability;
+      label = object_classification.label;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (scepter914) [reply common/perception_utils/include/perception_utils/object_classification.hpp:37]: To prevent bigger PR, I fix another PR.
So I add TODO at [bee5d5d](https://github.com/autowarefoundation/autoware.universe/pull/1954/commits/bee5d5d5175ffeefef0a7c5462a4771642434388).

File: common/perception_utils/test/src/test_matching.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+{
+geometry_msgs::msg::Pose createPose(const double x, const double y, const double yaw)
+{
+  geometry_msgs::msg::Pose p;
+  p.position.x = x;
+  p.position.y = y;
+  p.position.z = 0.0;
+  p.orientation = tier4_autoware_utils::createQuaternionFromYaw(yaw);
+
+  return p;
```
[the remark this answers is not in the document]
Comment (scepter914) [reply common/perception_utils/test/src/test_matching.cpp:34]: Thank you for useful information.
I refactored at [517b97f](https://github.com/autowarefoundation/autoware.universe/pull/1954/commits/517b97f52c1ff325dd9b8521fa78f60ee64a8745).

[the remark this answers is not in the document]
Comment (kenji-miyake) [reply common/perception_utils/test/src/test_matching.cpp:34]: Thanks.
I think it can also be applied to `geometry_msgs::msg::Pose`, but I'll leave it to you.

File: common/perception_utils/include/perception_utils/transform.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (scepter914) [reply common/perception_utils/include/perception_utils/transform.hpp:35]: Unfortunately, build is failed without `[[maybe_unused]]` in my environment.
So I would like to fix this problem in another PR.

```
/install/perception_utils/include/perception_utils/transform.hpp:35:48: error: ‘boost::optional<geometry_msgs::msg::Transform_<std::allocator<void> > > {anonymous}::getTransform(const tf2_ros::Buffer&, const string&, const string&, const rclcpp::Time&)’ defined but not used [-Werror=unused-function]
   35 | boost::optional<geometry_msgs::msg::Transform> getTransform(
```

File: common/perception_utils/include/perception_utils/transform.hpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+#else
+#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+#endif
+
+#include "boost/optional.hpp"
+
+#include <string>
+
+namespace
+{
```
Comment (kenji-miyake) [inline common/perception_utils/include/perception_utils/transform.hpp:34]: I think unnamed namespaces in a header file should be avoided.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf21-dont-use-an-unnamed-anonymous-namespace-in-a-header

Comment (scepter914) [reply common/perception_utils/include/perception_utils/transform.hpp:34]: Thank you for helpful review.
As same as other fixing point, I would like to fix this point to prevent bigger changing point.

**Comment by scepter914** [comment]
memo: TODO list
- <https://github.com/autowarefoundation/autoware.universe/pull/1954#discussion_r985329040> fix unnamed namespace
- <https://github.com/autowarefoundation/autoware.universe/pull/1954#discussion_r979912824> fix unused flag
- <https://github.com/autowarefoundation/autoware.universe/pull/1954#discussion_r979909631> change to STL

**Comment by scepter914** [comment]
@miursh cc. @Shin-kyoto 
This PR would conflict to [the PR for distance threshold list](https://github.com/autowarefoundation/autoware.universe/pull/1902).
As concerned by perception_util architecture, it might be better to merge this PR earlier than it.

**Comment by scepter914** [comment]
@miursh 
fix conflict for https://github.com/autowarefoundation/autoware.universe/pull/1902
Would you review please?

**Review by miursh — APPROVED** [review]
LGTM
I confirmed it's working as same as before.
